cuda/matrix/ell_kernels.cu

```diff
@@ -151,17 +261,98 @@ void advanced_spmv(std::shared_ptr<const CudaExecutor> exec,
                    const matrix::Dense<ValueType> *beta,
                    matrix::Dense<ValueType> *c)
 {
+    int subwarp_size = 1;
+    const auto nrows = a->get_size()[0];
+    const auto ell_ncols = a->get_num_stored_elements_per_row();
+    bool atomic = false;
+    int nwarps_per_row = 1;
+    const auto nwarps = exec->get_num_cores_per_sm() / cuda_config::warp_size *
+                        exec->get_num_multiprocessor() * multiple;
+    if (static_cast<double>(ell_ncols) / nrows > ratio) {
+        while (subwarp_size < 32 && (subwarp_size << 1) <= ell_ncols) {
```

Comment:
I think an explaining comment would not hurt here...

Reply:
done